crepecake/src/main/java/com/google/cloud/tools/crepecake/blob/BlobDescriptor.java

```diff
@@ -67,18 +78,18 @@ public long getSize() {
    * Two {@link BlobDescriptor} objects are equal if their
    *
    * <ol>
-   *   <li>{@code digest}s are not null and equal
-   *   <li>{@code size} is equal
+   *   <li>{@code digest}s are not null and equal, and
+   *   <li>{@code size}s are non-negative and equal
    * </ol>
    */
   @Override
   public boolean equals(Object obj) {
-    if (digest == null || !(obj instanceof BlobDescriptor)) {
+    if (digest == null || size < 0 || !(obj instanceof BlobDescriptor)) {
```

Comment:
Given the weird logic here when size or digest is missing, I think you need to first check for identity. I.e.

if (obj == this) return true;

Otherwise this violates the contract of equals.